storage-node-v2/src/services/sync/storageObligations.ts

```diff
@@ -0,0 +1,215 @@
+}
+
+/**
+ * Data object abstraction.
+ */
+type DataObject = {
+  /**
+   * Data object ID
+   */
+  id: string
+
+  /**
+   * Assigned bag ID
+   */
+  bagId: string
+}
+
+/**
+ * Get storage provider obligations like (assigned data objects) from the
+ * runtime (Query Node).
+ *
+ * @param queryNodeUrl - Query Node URL
+ * @param workerId - worker ID
+ * @returns promise for the DataObligations
+ */
+export async function getStorageObligationsFromRuntime(
+  queryNodeUrl: string,
+  workerId: number
+): Promise<DataObligations> {
+  const api = new QueryNodeApi(queryNodeUrl)
+
+  const allBuckets = await getAllBuckets(api)
+
+  const bucketIds = allBuckets
+    .filter((bucket) => bucket.operatorStatus?.workerId === workerId)
+    .map((bucket) => bucket.id)
+  const assignedBags = await getAllAssignedBags(api, bucketIds)
+
+  const bagIds = assignedBags.map((bag) => bag.id)
+  const assignedDataObjects = await getAllAssignedDataObjects(api, bagIds)
+
+  const model: DataObligations = {
+    storageBuckets: allBuckets.map((bucket) => ({
+      id: bucket.id,
+      operatorUrl: bucket.operatorMetadata?.nodeEndpoint ?? '',
+      workerId: bucket.operatorStatus?.workerId,
+    })),
+    bags: assignedBags.map((bag) => ({
+      id: bag.id,
+      buckets: bag.storageBuckets.map((bucketInBag) => bucketInBag.id),
+    })),
+    dataObjects: assignedDataObjects.map((dataObject) => ({
+      id: dataObject.id,
+      bagId: dataObject.storageBagId,
+    })),
+  }
+
+  return model
+}
+
+/**
+ * Get storage bucket IDs assigned to the worker.
+ *
+ * @param queryNodeUrl - Query Node URL
+ * @param workerId - worker ID
+ * @returns storage bucket IDs
+ */
+export async function getStorageBucketIdsByWorkerId(queryNodeUrl: string, workerId: number): Promise<string[]> {
+  const api = new QueryNodeApi(queryNodeUrl)
+
+  const idFragments = await api.getStorageBucketIdsByWorkerId(workerId.toString())
+  const ids = idFragments.map((frag) => frag.id)
+
+  return ids
+}
+
+/**
+ * Get IDs of the data objects assigned to the bag ID.
+ *
+ * @param api - initialiazed QueryNodeApi instance
+ * @param bagId - bag ID
+ * @returns data object IDs
+ */
+export async function getDataObjectIDsByBagId(queryNodeUrl: string, bagId: string): Promise<string[]> {
+  const api = new QueryNodeApi(queryNodeUrl)
+  const dataObjects = await getAllAssignedDataObjects(api, [bagId])
+
+  return dataObjects.map((obj) => obj.id)
+}
+
+/**
+ * Get all storage buckets registered in the runtime (Query Node).
+ *
+ * @param api - initialiazed QueryNodeApi instance
+ * @returns storage buckets data
+ */
+async function getAllBuckets(api: QueryNodeApi): Promise<StorageBucketDetailsFragment[]> {
+  const idFragments = await api.getStorageBucketIds()
+  const ids = idFragments.map((frag) => frag.id)
+
+  return await getAllObjectsWithPaging(
+    'get all storage buckets',
+    async (offset, limit) => await api.getStorageBucketDetails(ids.slice(offset, offset + limit), 0, limit)
```

Comment:
Currently 2 queries are executed, even if the number of buckets is < 1000:
```
colossus-1               | 2021-11-30 13:00:28:028 debug: Sync - getting get all storage buckets: offset = 0, limit = 1000
colossus-1               | 2021-11-30 13:00:28:028 debug: Sync - getting get all storage buckets: offset = 1000, limit = 1000
```
To optimize that I suggest changing `getAllObjectsWithPaging` function in a following way:
```typescript
async function getAllObjectsWithPaging<T>(
  objectName: string,
  query: (offset: number, limit: number) => Promise<T[]> | false
)
```
Allowing query to return either `Promise<T[]>` or `false`. In case `false` is returned - the function should terminate the loop.

This should allow for something like:
```suggestion
    (offset, limit) => ids.slice(offset, offset + limit).length ? api.getStorageBucketDetails(ids.slice(offset, offset + limit), 0, limit) : false
```

Reply:
Fixed here: https://github.com/Joystream/joystream/pull/2686/commits/bcb65423693a3e0f2831aa60271e06d7524cd7bf